prov/tcp/src/tcpx_cq.c

```diff
@@ -161,16 +161,13 @@ void tcpx_cq_report_success(struct util_cq *cq,
 		return;
 
 	flags = xfer_entry->flags & ~TCPX_INTERNAL_MASK;
-	if (flags & FI_RECV) {
+	if (flags & FI_RECV)
 		len = xfer_entry->hdr.base_hdr.size -
 		      xfer_entry->hdr.base_hdr.hdr_size;
-		tcpx_get_cq_info(xfer_entry, &flags, &data, &tag);
-	} else {
+	else
 		len = 0;
-		data = 0;
-		tag = 0;
```

Comment:
tcp sets FI_REMOTE_WRITE in tcpx_op_write() as part of receive side RMA write handling.  I *think* it's not setting FI_REMOTE CQ_DATA flag.  I'll verify if this is the case.

Reply:
See if #7136 applied before this series fixes the problem.  I don't see where tcp was setting FI_REMOTE_CQ_DATA as part of the completion.  The original change Juee made ended up adding that through the tcpx_get_cq_info() call.

Reply:
It didn't help :(
The check above that I added shows that the flags reported that there is CQ data but the CQ data didn't match so I think something else is happening maybe but I'm not quite sure what. My gdb step through looks like the failure is happening on the side that is sending CQ data but shouldn't be receiving the CQ data. The receiving side actually passes the validation. Maybe a reuse of an entry somewhere? I tried resetting the flags and also doing a memset after every entry alloc (tcpx_alloc_xfer) but that didn't do anything. Can you think of any other potential missing memset?

Reply:
What I mean is that the side that sent all the writedatas is failing on the send completion because the send completion is adding the FI_REMOTE_CQ_DATA flag in when it shouldn't.
The receive side is fine. The FI_REMOTE_WRITE flag gets set which means it goes into the first clause of the if here and calls tcpx_get_cq_info which ORs in the CQ data flag [here](https://github.com/ofiwg/libfabric/blob/main/prov/tcp/src/tcpx_cq.c#L139)

The problem is that on the send side we also write the CQ data flag into the completion because we OR in the flags passed in from the call [here](https://github.com/ofiwg/libfabric/blob/main/prov/tcp/src/tcpx_rma.c#L235)

Removing the flags from the above setting fixes the issue and lets the test pass. I think there is a similar issue with sends as well since it also ORs in the msg flags [here](https://github.com/ofiwg/libfabric/blob/main/prov/tcp/src/tcpx_msg.c#L254) but I haven't verified. All of our CQ completion verification in fabtests (including these ubertest changes) has been on receive side completions so we haven't caught it.

Reply:
I'm suggesting that the fix needs to be re-examined.

This is the current code in tcpx_op_write():
```
if (ep->cur_rx.hdr.base_hdr.flags & TCPX_REMOTE_CQ_DATA)
    rx_entry->flags = (FI_COMPLETION | FI_REMOTE_WRITE);
```
It's checking that the RMA write carries remote CQ data, but then doesn't set that flag?  That *looks* wrong.  Instead, it sets the FI_REMOTE_WRITE flag, then expects some code down the line with check if FI_REMOTE_WRITE is set, and if so, recheck if TCPX_REMOTE_CQ_DATA has been set, so it can add in the FI_REMOTE_CQ_DATA flag.  That's convoluted.  We're already checking for remote CQ data here, just OR in the flag. That will at least make the if statement above look reasonable.

That's the first issue.  The second issue is that apparently the CQ data is never being reported.  That probably needs to be handled as separate cases in tcpx_cq_report_success() and tcpx_cq_report_error(), and leave tcpx_get_cq_info() as a receive only function.

Separately, tcpx_op_write() should probably be updated to only perform the check for TCPX_REMOTE_CQ_DATA once, rather than twice like it currently does (or 3 times if you include the call it makes to tcpx_validate_rx_rma_data).

Reply:
Gotcha. I thought you were suggesting 7136 as a complete fix as is. I completely agree. We should add 7136, remove the data part from the set cq info call so we keep it as an only recv function and separate the data into a separate set cq call that is triggered by FI_REMOTE_CQ_DATA

Reply:
#7136 updated with an untested fix to also set the CQ data.  Untested in that fabtests passes, but no idea if the fix works since we don't have a test for it outside of this commit.

Reply:
Btw, I *want* to set the cq data simply by checking for FI_REMOTE_CQ_DATA, but I think that flag is currently set for outbound transfers (that you referenced above).  I'm trying to figure out what to do on the send side with the flags.

Happy birthday.

Reply:
I've confirmed that #7136 fixes the receive side completion reporting with the fi_rma_bw test + validation. Juee is verifying with her ubertest changes but I think if that works, 7136 can be merged and the patch removed from this patch set.

Reply:
I have verified that #7136 works with my ubertest changes.